### PR TITLE
New function for cross sections (RHI) from radar volume

### PR DESCRIPTION
**Pull request by JulianGiles**
I have added a function (cross_section_ppi) that extracts a vertical cross section (like RHI) from a radar volume (collection of PPIs). I have also slightly modified the georeference_dataset function to process 3D volumes.

### CODE REVIEW

**Comment by JulianGiles** [comment]
I will fix the style in a bit

**Comment by kmuehlbauer** [comment]
@JulianGiles You would need to run `black` on the changed files to fix the linting error. Unfortunately I can't push the needed changes since you are using the main-branch on your fork.

File: wradlib/util.py [hunk 1]
```diff
@@ -987,3 +989,233 @@ def has_import(module):
 
 if __name__ == "__main__":
     print("wradlib: Calling module <util> as main...")
+
+
+def vertical_interpolation(vol, elevs=None, method="nearest"):
+    """
+    Vertically interpolate volume data
+
```
Comment (kmuehlbauer) [inline wradlib/util.py:997]: Please add proper `Parameters` and `Keyword Arguments` in the docstring.

Maybe change function name to `vertical_interpolate_volume`.

File: wradlib/util.py [hunk 2]
```diff
@@ -987,3 +989,233 @@ def has_import(module):
...
+        arbitrary points in the x,y space of the georeferenced object: [ (x1, y1), (x2,y2) ].
+        In case two points are given, a cross section along the line connecting
+        the points will be generated by selecting the nearest-neighbor values of data.
+        No interpolation of data is performed. If more than two points are given, only
+        the first two are used. The resulting dataset has dimensions xyi (which is just
+        an index along the line connecting the points) and elevation, and
+        coordinates xy (distance along the line from p1) and z. The xy and z coordinates
+        should be used for plotting.
+
+    #######################
```
Comment (kmuehlbauer) [inline wradlib/util.py:1039]: remove ###-line

File: wradlib/util.py [hunk 3]
```diff
@@ -987,3 +989,233 @@ def has_import(module):
...
+    proj : :py:class:`gdal:osgeo.osr.SpatialReference`, :py:class:`cartopy.crs.CRS` or None
+        Projection to use with :py:class:`wradlib.georef.xarray.georeference_dataset`.
+        If GDAL OSR SRS, output is in this projection, else AEQD.
+    Returns
+    ----------
+    obj : :py:class:`xarray:xarray.Dataset` or :py:class:`xarray:xarray.DataArray`
+        Dataset of cross section(s) in the specified azimuth(s) or along the line
+        connecting the given points.
+    """
+    import xarray as xr
```
Comment (kmuehlbauer) [inline wradlib/util.py:1065]: move all imports to the top of file

File: wradlib/util.py [hunk 4]
```diff
@@ -987,3 +989,233 @@ def has_import(module):
...
+            # if some of the points is outside the radar volume area raise an exception
+            test = np.array(
+                [~(ds.x.min() < x1 < ds.x.max()),
+                 ~(ds.x.min() < x2 < ds.x.max()),
+                 ~(ds.y.min() < y1 < ds.y.max()),
+                 ~(ds.y.min() < y2 < ds.y.max())]
+            )
+
+            if test.any():
+                raise Exception("At least one of the points given is outside of the radar volume area")
```
Comment (kmuehlbauer) [inline wradlib/util.py:1159]: That sounds like we should raise `ValueError` here.

File: wradlib/util.py [hunk 5]
```diff
@@ -987,3 +989,233 @@ def has_import(module):
...
+
+        except TypeError:
+            # `azimuth` is not a list of azimuths nor a couple of points
+            raise Exception("Not azimuth values nor points was provided to `azimuth`")
```
Comment (kmuehlbauer) [inline wradlib/util.py:1163]: Re-raise as TypeError here

File: wradlib/util.py [hunk 6]
```diff
@@ -987,3 +989,233 @@ def has_import(module):
...
+
+        # Check that the two points given are not the same
+        if (p1 == p2).all():
+            raise Exception("p1=p2. The two points given are the same. Please give different points.")
```
Comment (kmuehlbauer) [inline wradlib/util.py:1167]: Raise ValueError here

File: wradlib/util.py [hunk 7]
```diff
@@ -987,3 +989,233 @@ def has_import(module):
...
+
+        # number of points to make the line between p1 and p2 (should be greater
+        # than the resolution of the volume)
+        nn = 1000
```
Comment (kmuehlbauer) [inline wradlib/util.py:1171]: Should we parameterize this, as kwarg?

File: wradlib/georef/xarray.py [hunk 8]
```diff
@@ -210,6 +210,10 @@ def georeference_dataset(obj, **kwargs):
         dimlist = list(obj["elevation"].dims)
     else:
         dimlist = list(obj["azimuth"].dims)
+
+    if xyz.ndim > 3:
```
Comment (kmuehlbauer) [inline wradlib/georef/xarray.py:214]: Are other Datasets/DataArrays possible which have more than 3 dimensions where this doesn't fit? We might need to check the dimensions here.

**Review by kmuehlbauer — CHANGES_REQUESTED** [review]
@JulianGiles Thanks for tackling this most wanted enhancement. I've added some comments.

File: wradlib/georef/xarray.py (hunk 8 above)
Comment (JulianGiles) [reply wradlib/georef/xarray.py:214]: xyz is an array of cartesian coordinates for every spherical coordinate, so the possible dimensions are: elevation, azimuth, range, 3. For 2d, it either has (elevation, range, 3) or (azimuth, range, 3) dimensions. For 3d, the only option is the full (elevation, azimuth, range, 3) dimensions. Thus, i think adding these lines for the 3d case will not break other functionalities

Comment (kmuehlbauer) [reply wradlib/georef/xarray.py:214]: > xyz is an array of cartesian coordinates for every spherical coordinate, so the possible dimensions are: elevation, azimuth, range, 3. For 2d, it either has (elevation, range, 3) or (azimuth, range, 3) dimensions. For 3d, the only option is the full (elevation, azimuth, range, 3) dimensions. 

Ah, yes, great! Please add your above quoted explanation above the if-clause as comment for further reference. Thanks!

File: wradlib/util.py [hunk 9]
```diff
@@ -987,3 +991,300 @@ def has_import(module):
...
+    -----------------
+    elevs : iterable of elevations to which interpolate the data. Defaults to None, which does no interpolation and returns a stacked array of the data.
+    method : method for interpolation, defaults to "nearest"
+
+    Returns
+    ----------
+    ds : :py:class:`xarray:xarray.Dataset`
+
+    """
+    import xarray as xr
```
Comment (kmuehlbauer) [inline wradlib/util.py:1014]: Remove unneeded import here.

**Comment by kmuehlbauer** [comment]
@JulianGiles The notebook tests failures are unrelated.

**Comment by JulianGiles** [comment]
How can I provide you some tests? Can I send some lines of codes to you by email? Also what do you need in order to add an entry about the function in the docs?

**Comment by kmuehlbauer** [comment]
- The entry in the docs is generated automatically via autodoc/sphinx, no action needed from your side.
- Tests are added in the relevant file below https://github.com/wradlib/wradlib/tree/main/wradlib/tests. We use `pytest` for testing. Please let me know if you need more information.

**Comment by JulianGiles** [comment]
Ok, I will try to add a test on test_util.py.
For the docs, I meant if I can provide you with a notebook to put an example (in addition to the auto-generated entry), something like https://docs.wradlib.org/en/stable/notebooks/workflow/recipe2.html

**Comment by kmuehlbauer** [comment]
@JulianGiles You can add the tests to this PR for sure.

**Comment by kmuehlbauer** [comment]
@JulianGiles Yes, we can add that notebook to [wradlib-notebooks](https://github.com/wradlib/wradlib-notebooks). That's the place where the tutorials and examples are hosted.

**Comment by kmuehlbauer** [comment]
Thanks @JulianGiles for trying and for the example notebook. I'll take the steps from here now, to get everything in place.

**Comment by kmuehlbauer** [comment]
I'm going to merge this now as is. We can fix anything that pops up when adding tests.

**Comment by kmuehlbauer** [comment]
:tada: Great enhancement, congratulations @JulianGiles

**Pull request merged**